### PR TITLE
[Feature] Added support for vertex skinning

### PR DESCRIPTION
**Pull request by HedvigAxelsson**
### Purpose of this PR
Added support for vertex skinning.

---
### Release Notes

---
### Testing status
Tested in A2 project.

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=master&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Low

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs [hunk 1]
```diff
@@ -291,21 +291,14 @@ internal void RenderMirrorView(CommandBuffer cmd)
             using (new ProfilingSample(cmd, "XR Mirror View"))
             {
                 cmd.SetRenderTarget(BuiltinRenderTextureType.CameraTarget);
-
-            #if UNITY_2020_1_OR_NEWER
```
Comment (phi-lira) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs]: why was this removed? This prevents us from backporting this PR to 19.3 and ship in a package for Unite.

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs [hunk 2]
```diff
@@ -35,6 +38,10 @@ internal struct Attributes
             Vector4 uv3;
             [Semantic("COLOR")][Optional]
             Vector4 color;
+            [Semantic("BLENDWEIGHTS")][Optional]
+            Vector4 weights;
+            [Semantic("BLENDINDICES")][Optional]
+            UInt32_4 indicies;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: indices

File: com.unity.render-pipelines.high-definition/Editor/ShaderGraph/HDSubShaderUtilities.cs [hunk 3]
```diff
@@ -36,6 +39,8 @@ internal struct AttributesMesh
             [Semantic("TEXCOORD1")][Optional]       Vector4 uv1;
             [Semantic("TEXCOORD2")][Optional]       Vector4 uv2;
             [Semantic("TEXCOORD3")][Optional]       Vector4 uv3;
+            [Semantic("BLENDWEIGHTS")][Optional]    Vector4 weights;
+            [Semantic("BLENDINDICES")][Optional]    UInt32_4 indicies;
```
Comment (phi-lira) [inline com.unity.render-pipelines.high-definition/Editor/ShaderGraph/HDSubShaderUtilities.cs]: indices

File: com.unity.render-pipelines.high-definition/Editor/ShaderGraph/HDSubShaderUtilities.cs [hunk 4]
```diff
@@ -271,6 +278,9 @@ internal struct VertexDescriptionInputs
                 new Dependency("VertexDescriptionInputs.uv2",                       "AttributesMesh.uv2"),
                 new Dependency("VertexDescriptionInputs.uv3",                       "AttributesMesh.uv3"),
                 new Dependency("VertexDescriptionInputs.VertexColor",               "AttributesMesh.color"),
+
+                new Dependency("VertexDescriptionInputs.BoneWeights",               "AttributesMesh.weights"),
+                new Dependency("VertexDescriptionInputs.BoneIndices",               "AttributesMesh.indicies")
```
Comment (phi-lira) [inline com.unity.render-pipelines.high-definition/Editor/ShaderGraph/HDSubShaderUtilities.cs]: indices

File: com.unity.render-pipelines.high-definition/Editor/ShaderGraph/VertexAnimation.template.hlsl [hunk 5]
```diff
@@ -32,6 +32,8 @@ VertexDescriptionInputs AttributesMeshToVertexDescriptionInputs(AttributesMesh i
     $VertexDescriptionInputs.uv2:                       output.uv2 =                         input.uv2;
     $VertexDescriptionInputs.uv3:                       output.uv3 =                         input.uv3;
     $VertexDescriptionInputs.VertexColor:               output.VertexColor =                 input.color;
+    $VertexDescriptionInputs.BoneWeights:               output.BoneWeights =                 input.weights;
+    $VertexDescriptionInputs.BoneIndices:               output.BoneIndices =                 input.indicies;
```
Comment (phi-lira) [inline com.unity.render-pipelines.high-definition/Editor/ShaderGraph/VertexAnimation.template.hlsl]: indices

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/BuildVertexDescriptionInputs.template.hlsl [hunk 6]
```diff
@@ -32,6 +32,8 @@ VertexDescriptionInputs BuildVertexDescriptionInputs(Attributes input)
     $VertexDescriptionInputs.uv3:                       output.uv3 =                         input.uv3;
     $VertexDescriptionInputs.VertexColor:               output.VertexColor =                 input.color;
     $VertexDescriptionInputs.TimeParameters:            output.TimeParameters =              _TimeParameters.xyz;
+    $VertexDescriptionInputs.BoneWeights:               output.BoneWeights =                 input.weights;
+    $VertexDescriptionInputs.BoneIndices:               output.BoneIndices =                 input.indicies;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/Templates/BuildVertexDescriptionInputs.template.hlsl]: indices

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs [hunk 7]
```diff
@@ -16,7 +16,10 @@ static class UniversalShaderGraphResources
         public static string s_ResourceClassName => typeof(UniversalShaderGraphResources).FullName;
 
         public static string s_AssemblyName => typeof(UniversalShaderGraphResources).Assembly.FullName.ToString();
-        
+
+        struct UInt32_4
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: I dont' understand why an empty struct. How do we access the indices?

Comment (HedvigAxelsson) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: The struct is for the type declaration for the shader. We access them in the shader.

Comment (HedvigAxelsson) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: It's used in ShaderSpliceUtils.cs, function GetUintCount(string typeName)

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs [hunk 8]
```diff
@@ -291,7 +291,7 @@ internal void RenderMirrorView(CommandBuffer cmd)
             using (new ProfilingSample(cmd, "XR Mirror View"))
             {
                 cmd.SetRenderTarget(BuiltinRenderTextureType.CameraTarget);
-
+
```
Comment (phi-lira) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/XR/XRSystem.cs]: minor formatting issue.

File: com.unity.shadergraph/Editor/Data/Graphs/ShaderGraphRequirements.cs [hunk 9]
```diff
@@ -104,6 +105,12 @@ public bool requiresTime
             internal set { m_RequiresTime = value; }
         }
 
+        public bool requiresVertexSkinning
```
[the remark this answers is not in the document]
Comment (HedvigAxelsson) [reply com.unity.shadergraph/Editor/Data/Graphs/ShaderGraphRequirements.cs]: Test status is same as master: i.e everything but standalone works. Gonna rerun tests with your suggestions.

**Comment by alindmanUnity** [comment]
just leaving a quick note that we have a branch that changes the generation process for shaders from shader graph quite a lot, and we'll get some merge conflicts with this change. we need to coordinate landing pattern c:

**Comment by HedvigAxelsson** [comment]
@alindmanUnity Sure, let's talk about landing pattern. Do you have any ideas? I'm assuming the new node is the main pain point here, but I guess it should be easy to just rewrite it to the new generation process?
